subscription/src/main/java/org/killbill/billing/subscription/engine/dao/DefaultSubscriptionDao.java

```diff
@@ -874,17 +875,20 @@ private List<DefaultSubscriptionBase> buildBundleSubscriptions(final List<Defaul
                     }
 
                     if (baseTriggerEventForAddOnCancellation != null) {
-                        final SubscriptionBaseEvent addOnCancelEvent = new ApiEventCancel(new ApiEventBuilder()
-                                                                                                  .setSubscriptionId(reloaded.getId())
-                                                                                                  .setEffectiveDate(baseTriggerEventForAddOnCancellation.getEffectiveDate())
-                                                                                                  .setCreatedDate(baseTriggerEventForAddOnCancellation.getCreatedDate())
-                                                                                                  // This event is only there to indicate the ADD_ON is future canceled, but it is not there
-                                                                                                  // on disk until the base plan cancellation becomes effective
-                                                                                                  .setFromDisk(false));
-
+                        
+                        ApiEventBuilder builder = new ApiEventBuilder()
+                                .setSubscriptionId(reloaded.getId())
+                                .setEffectiveDate(baseTriggerEventForAddOnCancellation.getEffectiveDate())
+                                .setCreatedDate(baseTriggerEventForAddOnCancellation.getCreatedDate())
+                                // This event is only there to indicate the ADD_ON is future canceled, but it is not there
+                                // on disk until the base plan cancellation becomes effective
+                                .setFromDisk(false);
+                        builder = cur.getIncludeDeletedEvents() ? builder.setActive(baseTriggerEventForAddOnCancellation.isActive()) : builder;
```

Comment:
I am wondering if we really need this logic: We are creating an in memory cancel event for the add-on as a result of seeing a future cancelation for the base plan, so I would think such computed event always need to be active. Was there a reason (test failing?) why you added this?

Reply:
I added this to prevent creating an active transition corresponding to a deleted event.

See this [test](https://github.com/reshmabidikar/killbill/blob/835a6a5cf9344da4effd058bd05167f6b6301794/beatrix/src/test/java/org/killbill/billing/beatrix/integration/TestSubscription.java#L1014). When the base subscription is uncancelled [here](https://github.com/reshmabidikar/killbill/blob/835a6a5cf9344da4effd058bd05167f6b6301794/beatrix/src/test/java/org/killbill/billing/beatrix/integration/TestSubscription.java#L1078), the base cancel event is deleted. However, when `includeDeleted=true` is passed, this returns the deleted cancel event and results in the `baseTriggerEventForAddOnCancellation` getting initialized [here](https://github.com/reshmabidikar/killbill/blob/835a6a5cf9344da4effd058bd05167f6b6301794/subscription/src/main/java/org/killbill/billing/subscription/engine/dao/DefaultSubscriptionDao.java#L844). This in turn results in a cancel event getting created for the addon. By default, it is created with `isActive=true` which results in it being used to create an active transition [here](https://github.com/reshmabidikar/killbill/blob/835a6a5cf9344da4effd058bd05167f6b6301794/subscription/src/main/java/org/killbill/billing/subscription/api/user/DefaultSubscriptionBase.java#L845). Hence it needs to be set with the `isActive` value in the `baseTriggerEventForAddOnCancellation`.

Reply:
I see what you are saying, and what you have is probably OK, but I think we could simplify even a bit more: Regardless of the mode, `includeDeletedEvents=true|false`, I think we should only add a new in-memory event in the stream for the add-on when there is an active event. So instead of just looking at whether `baseTriggerEventForAddOnCancellation` is not null we should check whether `baseTriggerEventForAddOnCancellation` is not null and is active [here](https://github.com/reshmabidikar/killbill/blob/work-for-release-0.23x-1030-blocking-states/subscription/src/main/java/org/killbill/billing/subscription/engine/dao/DefaultSubscriptionDao.java#L877).

Reply:
Fixed